AIX support

wiggin15 opened the pull request:
Following #605, this is the proposed branch for AIX support.

The following methods and methods are not supported in AIX in this branch:

    psutil.Process.memory_maps
    psutil.Process.num_ctx_switches

Not all unit tests pass on AIX at this time (I didn't make any changes to unit tests yet).

giampaolo commented on psutil/TODO.aix:
```diff
@@ -0,0 +1,30 @@
+Process.io_counters read count is always 0
+
+
+TestSystemAPIs.test_pid_exists_2           there are pids in /proc that don't really exist
```
You mean, say, a `/proc/54` directory exists but there's no such PID?

wiggin15 replied:
Yes, AIX is weird like that.

giampaolo replied:
That's *really* unfortunate. How about doing this then?

```
def pids():
    """Returns a list of PIDs currently running on the system."""
    return [int(x) for x in os.listdir('/proc') if x.isdigit() \
            and  _psposix.pid_exists(int(x))]
```

Does this test pass?

wiggin15 replied:
This fixes `test_pid_exists_2` but causes `test_pids` to break because `ps` still returns the processes that "don't really exist" and then we get this:

      File "/root/psutil/psutil/tests/test_posix.py", line 297, in test_pids
         self.fail("difference: " + str(difference))
    AssertionError: difference: [131076, 196614, 1048608, 1114146, 1179684, 1376298, 1638568]

giampaolo commented on psutil/TODO.aix:
```diff
@@ -0,0 +1,30 @@
+
+
+TestSystemAPIs.test_pid_exists_2           there are pids in /proc that don't really exist
+TestProcess.test_name                      isolated python calls execve which changes process name
```
What does this mean?

wiggin15 replied:
On our environment we're using a custom version of Python, which on AIX uses a trick utilizing execve. This means the test will work on regular environments but not on ours specifically. I'm not sure this file should really be included in the repo (I'm also not sure it's up-to-date), this file was created for note taking when debugging the tests.

giampaolo replied:
I'm not sure what you mean by this then. Simply make sure `name()` returns the expected string (in case of a Python process it should be "python".

wiggin15 replied:
execve causes the name to change so `sys.executable` and `name` don't match. 

    AssertionError: ('python2.7', 'python2.7.bin')

This is implementation detail in our environment and can be ignored.
I found out that the set of tests that fail is very different from before (it's more stable now too, i.e. less "flaky" tests). This file will be updated in the next commit.

giampaolo commented on psutil/TODO.aix:
```diff
@@ -0,0 +1,30 @@
+
+TestSystemAPIs.test_pid_exists_2           there are pids in /proc that don't really exist
+TestProcess.test_name                      isolated python calls execve which changes process name
+TestProcess.test_num_fds                   opening a socket doesn't create fd in /proc/pid/fd (until data is sent??)
```
Does this mean `num_fds()` is unreliable?

wiggin15 replied:
Hmm, I guess so

wiggin15 replied:
Apparently this doesn't fail any more

giampaolo commented on psutil/TODO.aix:
```diff
@@ -0,0 +1,30 @@
+TestSystemAPIs.test_pid_exists_2           there are pids in /proc that don't really exist
+TestProcess.test_name                      isolated python calls execve which changes process name
+TestProcess.test_num_fds                   opening a socket doesn't create fd in /proc/pid/fd (until data is sent??)
+TestProcess.test_open_files                /dev/null shows in open_files but it isn't a file
```
Can't you just skip it?

wiggin15 replied:
Yes. There are probably many changes to the tests I can do to make them pass. I deliberately did not change the unit tests for now to see what breaks - but I should before this will be really ready to be merged.

giampaolo commented on psutil/TODO.aix:
```diff
@@ -0,0 +1,30 @@
+TestProcess.test_name                      isolated python calls execve which changes process name
+TestProcess.test_num_fds                   opening a socket doesn't create fd in /proc/pid/fd (until data is sent??)
+TestProcess.test_open_files                /dev/null shows in open_files but it isn't a file
+TestProcess.test_pid_0                     pid 0 doesn't have a name on AIX
```
Does `ps` list it? If not is there some other task manager on AIX that does? If not I would advice to return `"kernel"`.

wiggin15 replied:
"ps" calls PID 0 "swapper". I can return this name specifically for the case of pid=0.

giampaolo replied:
Sounds good. Also make sure `pids()` return PID 0 and `pid_exists(0)` return True. If they don't by default then you can simply hard-code it. `ps` should be used as the leading reference and psutil should behave the same as ps.

giampaolo commented on psutil/TODO.aix:
```diff
@@ -0,0 +1,30 @@
+TestProcess.test_open_files                /dev/null shows in open_files but it isn't a file
+TestProcess.test_pid_0                     pid 0 doesn't have a name on AIX
+TestProcess.test_prog_w_funky_name         funky name test doesn't work on UNIX environments
+TestProcess.test_zombie_process            trying to create a zombie process doesn't create a zombie process on AIX
```
Do zombie processes exist on AIX? Is `status()` able to detect them?

wiggin15 replied:
AFAIK they do exist and can be detected, but the trick we use to create one with `fork` doesn't work

giampaolo replied:
What happens when you call this test function?
https://github.com/giampaolo/psutil/blob/8b8da39e0c62432504fb5f67c418715aad35b291/psutil/tests/__init__.py#L335
Does it crash? Does it return a process but it's status is not zombie (what is it?)?

wiggin15 replied:
This creates a process, but its status is 0 (SNONE) and not SZOMB. Then we crash here:

    call_until(lambda: zproc.status(), "ret == psutil.STATUS_ZOMBIE")

I'm not sure what that means.

giampaolo commented on psutil/TODO.aix:
```diff
@@ -0,0 +1,30 @@
+TestFetchAllProcesses.test_fetch_all       unknown
+TestProcess.test_Popen                     unknown (flaky)
+TestProcess.test_children                  unknown (flaky)
+TestProcess.test_children_recursive        unknown (flaky)
```
`children()` is a crucial method so I would like to see these pass. Since `children()` really is a utility method at its core, if they don't it's likely that `ppid()` implementation is broken.

wiggin15 replied:
This test actually didn't fail on the last test runs I ran, I don't know why it failed and if this still happens. Like I said, this file may be a little out-of-date.

giampaolo commented on psutil/__init__.py:
```diff
@@ -82,6 +82,7 @@
 from ._common import POSIX  # NOQA
 from ._common import SUNOS
 from ._common import WINDOWS
+from ._common import AIX
```
Please sort all imports by name.

giampaolo commented on psutil/_psaix.py:
```diff
@@ -0,0 +1,488 @@
+from ._common import socktype_to_enum, sockfam_to_enum
+from ._common import NIC_DUPLEX_FULL, NIC_DUPLEX_HALF, NIC_DUPLEX_UNKNOWN
+from ._common import usage_percent
+from ._compat import PY3
```
Please sort all imports by name. Also one import per line (avoid commas)

giampaolo commented on psutil/_psaix.py:
```diff
@@ -0,0 +1,488 @@
+    avail = free * PAGE_SIZE
+    used = inuse * PAGE_SIZE
+    percent = usage_percent((total - avail), total, _round=1)
+    return svmem(total, avail, percent, used, free)
```
This is one of the most important functions so there should be a specific test for this which tests the result against a command line utility (does AIX have `free`?)? As such there should be a `psutil/tests/test_aix.py` file.

wiggin15 replied:
No `free`, but a command called `svmem`. Will do.

wiggin15 replied:
I'm having trouble checking this function against the shell command because there memory usage changes between the call to `svmon` and the call to `psutil.virtual_memory`. I think checking whether they are "close" can be inconsistent. This function has basic tests in `test_system.py`...

giampaolo commented on psutil/_psaix.py:
```diff
@@ -0,0 +1,488 @@
+    if p.returncode != 0:
+        raise RuntimeError("%r command error\n%s" % (cmd, stderr))
+    processors = stdout.strip().splitlines()
+    return len(processors)
```
Looks like you want to add a test for this in `psutil/tests/test_aix.py`

wiggin15 replied:
I am already using a shell command to determine number of processors (lsdev). I'm not sure what I can test this against.

giampaolo replied:
You're right.

giampaolo replied:
So please just add a `# TODO: rewrite this in C` comment

wiggin15 replied:
I tried hard to find how to rewrite this in C and couldn't find any way to do that. Some functions I did find returned wrong results. I ran truss (the equivalent of strace) to find out how `lsdev` itself returns the results, and it looks like it iterates over the devices from the ODM (the equivalent of Windows registry). I don't think reading from the ODM in psutil is better than just parsing the output of `lsdev`.

wiggin15 replied:
Why return `None` from `cpu_count_physical` instead of 0?

giampaolo replied:
It's by design and documented: when cannot be determined we're supposed to return None. `os.cpu_count()` on Python 3 behaves the same.

giampaolo commented on psutil/_psaix.py:
```diff
@@ -0,0 +1,488 @@
+    """Return socket connections.  If pid == -1 return system-wide
+    connections (as opposed to connections opened by one process only).
+    """
+    cmap = _common.conn_tmap.copy()
```
Looks like there's no need to `copy()` the dict

wiggin15 replied:
This is the same implementation as in _pssunos.py

giampaolo commented on psutil/_psaix.py:
```diff
@@ -0,0 +1,488 @@
+        pass
+
+    def oneshot_exit(self):
+        pass
```
Looks like `cext.proc_basic_info()` returns multiple info so we can speed things up via `oneshot`().

giampaolo commented on psutil/_psaix.py:
```diff
@@ -0,0 +1,488 @@
+            # (no such process), no matter what.
+            # The process actually exists though, as it has a name,
+            # creation time, etc.
+            raise AccessDenied(self.pid, self._name)
```
I don't like the hard coded `(2, 3)` thing. Looks like you should use `pid_exists()` instead.

wiggin15 replied:
This is the same implementation as in _pssunos.py

giampaolo replied:
Then that should be changed but we can ignore it for now, If `pid_exists()` is reliable in this case (it should) then let's just use that.

giampaolo commented on psutil/_psaix.py:
```diff
@@ -0,0 +1,488 @@
+    @wrap_exceptions
+    def gids(self):
+        _, _, _, real, effective, saved = cext.proc_cred(self.pid)
+        return _common.puids(real, effective, saved)
```
These 2 (uids() and gids()) also look like something which can be sped up via `oneshot()`.

giampaolo commented on psutil/_psaix.py:
```diff
@@ -0,0 +1,488 @@
+        except OSError as err:
+            if err.errno == errno.ENOENT:
+                os.stat("/proc/%s" % self.pid)
+                return None
```
There's no precedent of `cwd()` returning None, but there is for `exe()` returning `""` so I would return an empty string here.

wiggin15 replied:
This is the same implementation as in _pssunos.py

giampaolo replied:
Oh I see. I don't remember why I did that but let's just return None for now.

giampaolo commented on psutil/_psaix.py:
```diff
@@ -0,0 +1,488 @@
+
+    @wrap_exceptions
+    def num_ctx_switches(self):
+        return _common.pctxsw(*cext.proc_num_ctx_switches(self.pid))
```
I thought this wasn't supported? (it's not from `psutil/__init__.py` at least)

wiggin15 replied:
Right. this should be removed.

giampaolo commented on psutil/_psutil_aix.c:
```diff
@@ -0,0 +1,777 @@
+    threadt = (perfstat_thread_t *)calloc(thread_count,
+        sizeof(perfstat_thread_t));
+    if (threadt == NULL) {
+        PyErr_SetFromErrno(PyExc_OSError);
```
You want `PyErr_NoMemory()` here.

giampaolo commented on psutil/_psutil_aix.c:
```diff
@@ -0,0 +1,777 @@
+static PyObject *
+psutil_users(PyObject *self, PyObject *args) {
+    struct utmpx *ut;
+    PyObject *ret_list = PyList_New(0);
```
In order to follow existing convention I recommend using a `py_` prefix for all python related variable names, so this should be `py_retlist`.

giampaolo commented on psutil/_psutil_aix.c:
```diff
@@ -0,0 +1,777 @@
+    struct utmpx *ut;
+    PyObject *ret_list = PyList_New(0);
+    PyObject *tuple = NULL;
+    PyObject *user_proc = NULL;
```
Same here (`py_tuple` and `py_user_proc`)

giampaolo commented on psutil/_psutil_aix.c:
```diff
@@ -0,0 +1,777 @@
+            "(sssfO)",
+            ut->ut_user,              // username
+            ut->ut_line,              // tty
+            ut->ut_host,              // hostname
```
Strings should be handled with care (it's a major change I introduced in 5.3.0). This is how they should be treated:
https://github.com/Infinidat/psutil/blob/c895a4aa1a08e7dac7213ab1c2458b0a28cea343/psutil/_psutil_linux.c#L469-L477

wiggin15 replied:
It looks like PyUnicode_DecodeFSDefault is available on Python 3 only? When using it my code doesn't compile. Looks like `#ifdef`s are required.

giampaolo replied:
I backported provided a "Python 2 like" version of `PyUnicode_DecodeFSDefault`  in `_psutil_common.c` so you just have to `#include "_psutil_common.h"` at the top of this C file in order to use it.

giampaolo commented on psutil/_psutil_aix.c:
```diff
@@ -0,0 +1,777 @@
+        py_tuple = Py_BuildValue(
+            "(ssss)",
+            mt->mnt_fsname,   // device
+            mt->mnt_dir,    // mount point
```
New way to handle strings:
https://github.com/Infinidat/psutil/blob/c895a4aa1a08e7dac7213ab1c2458b0a28cea343/psutil/_psutil_linux.c#L469-L477

giampaolo commented on psutil/_psutil_aix.c:
```diff
@@ -0,0 +1,777 @@
+    statp = (perfstat_netadapter_t *)
+        malloc(tot * sizeof(perfstat_netadapter_t));
+    if (statp == NULL) {
+        PyErr_SetFromErrno(PyExc_OSError);
```
Use `PyErr_NoMemory`

giampaolo commented on psutil/_psutil_aix.c:
```diff
@@ -0,0 +1,777 @@
+    /* allocate enough memory to hold the ncpu structures */
+    cpu = (perfstat_cpu_t *) malloc(ncpu * sizeof(perfstat_cpu_t));
+    if (cpu == NULL) {
+        PyErr_SetFromErrno(PyExc_OSError);
```
Use `PyErr_NoMemory`

giampaolo commented on psutil/_psutil_aix.c:
```diff
@@ -0,0 +1,777 @@
+error:
+    Py_XDECREF(py_cputime);
+    Py_DECREF(py_retlist);
+    free(cpu);
```
I think you should free() it only if it hasn't been malloc()'ed before. You can set `cpu = NULL;` earlier and do `if (cpu != NULL) {free(cpu);}` here

wiggin15 replied:
We can't get to `error` without allocating `cpu`, so this is safe.

giampaolo commented on psutil/_psutil_aix.c:
```diff
@@ -0,0 +1,777 @@
+    diskt = (perfstat_disk_t *)calloc(disk_count,
+        sizeof(perfstat_disk_t));
+    if (diskt == NULL) {
+        PyErr_SetFromErrno(PyExc_OSError);
```
PyErr_NoMemory

giampaolo commented on psutil/_psutil_aix.c:
```diff
@@ -0,0 +1,777 @@
+        );
+        if (py_disk_info == NULL)
+            goto error;
+        if (PyDict_SetItemString(py_retdict, diskt[i].name,
```
Correct way to handle strings:
https://github.com/Infinidat/psutil/blob/c895a4aa1a08e7dac7213ab1c2458b0a28cea343/psutil/_psutil_linux.c#L469-L477

giampaolo commented on psutil/_psutil_aix.h:
```diff
@@ -0,0 +1,25 @@
+/*
+ * Copyright (c) 2009, Giampaolo Rodola'. All rights reserved.
```
Please also add your name. Also do the same in other new files (copy license and add your name).

giampaolo commented on psutil/_psutil_posix.c:
```diff
@@ -688,7 +692,7 @@ void init_psutil_posix(void)
 #endif
 
-#if defined(PSUTIL_BSD) || defined(PSUTIL_OSX) || defined(PSUTIL_SUNOS)
+#if defined(PSUTIL_BSD) || defined(PSUTIL_OSX) || defined(PSUTIL_SUNOS) || defined(_AIX)
```
You should define `PSUTIL_AIX` in setup.py and use it in here

wiggin15 commented:
I added a new commit. Still no AIX-specific tests and no unicode string handling, but I fixed some broken functions and started changing the unit tests a bit. I think this looks better.

giampaolo commented on psutil/_psaix.py:
```diff
@@ -0,0 +1,525 @@
+        # an empty list means there were no connections for process or
+        # process is no longer active so we force NSP in case the PID
+        # is no longer there.
+        if not ret:
```
I suggest to remove the `if not ret` line. There may be some connections but the process may have disappeared beforehand so we basically want to always call `os.stat`

wiggin15 replied:
This is again the same as in _pssunos.py

giampaolo commented on psutil/_psaix.py:
```diff
@@ -0,0 +1,525 @@
+    @wrap_exceptions
+    def terminal(self):
+        psinfo = self._proc_basic_info()
+        ttydev = psinfo[-1]
```
I think you better use a "fixed" integer (5, 6 or whatever) because if you add a new element to the returned list this will break

giampaolo commented on psutil/_psaix.py:
```diff
@@ -0,0 +1,525 @@
+        if PY3:
+            stdout, stderr = [x.decode(sys.stdout.encoding)
+                              for x in (stdout, stderr)]
+        if "no such process" in stderr:
```
perhaps do `stderr.lower()`?

giampaolo commented on psutil/_psaix.py:
```diff
@@ -0,0 +1,525 @@
+            path = path.strip()
+            if path.startswith("//"):
+                path = path[1:]
+            if path == "Cannot be retrieved":
```
perhaps do `path.lower()`?

giampaolo commented on psutil/_psutil_aix.c:
```diff
@@ -0,0 +1,815 @@
+    Py_XDECREF(py_tuple);
+    Py_DECREF(py_retlist);
+    if (ut != NULL)
+        endutent();
```
You use `endutxent();` on line 288, not this. Why?

wiggin15 replied:
Looks like you're right.
Again, this is the same as in the sunos implementation.

giampaolo commented on psutil/_psutil_aix.c:
```diff
@@ -0,0 +1,815 @@
+        }
+    }
+    endutxent();
+    if (boot_time != 0.0) {
```
Are there cases where time can be 0.0? Perhaps we need a comment here?

wiggin15 replied:
Not sure. This too is copied from _psutil_sunos.c.

wiggin15 replied:
`boot_time` will be `0.0` if the loop breaks without finding `BOOT_TIME`. We can add a comment, but again this is common implementation with Solaris.

giampaolo replied:
Oh yeah. I now realize that sunos has different "minor" issues like this one (and others I mentioned here). I will stop being picky and just focus on more important aspects. =)

wiggin15 replied:
There are definitely some comments that are relevant and should be fixed in Solaris too. There are also a number of tests that fail in Solaris as well as on AIX that can be fixed in both. I'm keeping track of your comments and may submit a separate PR to fix them after this one.

giampaolo commented on psutil/_psutil_aix.c:
```diff
@@ -0,0 +1,815 @@
+        cpu.syscall,
+        cpu.devintrs,
+        cpu.softintrs,
+        cpu.traps
```
In TODO.aix you state all of these numbers are 0, correct?

wiggin15 replied:
Not all of them, only `syscall` and `traps`. I'm not sure why.

giampaolo replied:
OK then. If it's an OS problem there's nothing we can do.

giampaolo replied:
Please just add 2 XXX comments to signal that syscall and traps fields are always 0.

wiggin15 replied:
Ugh. My bad, there was a bug here (I don't like it when I have bugs :/). The values are ulonglong, so should be built with `K` instead of `I`. The values are _not_ 0.

giampaolo replied:
Perfect. Just update the doc/TODO.aix then

giampaolo commented on psutil/arch/aix/net_connections.c:
```diff
@@ -0,0 +1,328 @@
+            inet_ntop(fam, raddr, raddr_str, sizeof(raddr_str));
+            return Py_BuildValue("(iii(si)(si)ii)", fd, fam,
+                                 s.so_type, laddr_str, lport, raddr_str,
+                                 rport, state, pid);
```
I suspect returning 2 tuples in a single shot like this (`(si)`) will leak memory. You need 3 tuples, 2 per addresses and 1 per connection. It's tedious work but you can pretty much copy & paste from here:
https://github.com/giampaolo/psutil/blob/f435c2b6d308c12d0df33b10d828d97df303a614/psutil/_psutil_osx.c#L1332-L1349

giampaolo replied:
Note: you can use `make test-memleaks` to make sure of this.

wiggin15 replied:
What would be the result of `make test-memleaks` if there is a memory leak? Running this command returned 3 failed test:

    test_memory_leaks.TestProcessObjectLeaks.test_num_ctx_switches
    test_memory_leaks.TestTerminatedProcessLeaks.test_io_counters
    test_memory_leaks.TestTerminatedProcessLeaks.test_num_ctx_switches

The failures don't seem to be due to memory leaks.

giampaolo replied:
Let's think about memory leaks later, after you fixed unicode strings and /proc fs location.

giampaolo commented on psutil/arch/aix/net_connections.c:
```diff
@@ -0,0 +1,328 @@
+    msz = (size_t)(PROCSIZE * PROCINFO_INCR);
+    processes = (struct procentry64 *)malloc(msz);
+    if (!processes) {
+        PyErr_SetFromErrno(PyExc_OSError);
```
`PyErr_NoMemory()`

giampaolo commented on psutil/arch/aix/net_connections.c:
```diff
@@ -0,0 +1,328 @@
+            msz = (size_t)(PROCSIZE * (Np + PROCINFO_INCR));
+            processes = (struct procentry64 *)realloc((char *)processes, msz);
+            if (!processes) {
+                PyErr_SetFromErrno(PyExc_OSError);
```
`PyErr_NoMemory()`

giampaolo commented on psutil/arch/aix/net_connections.c:
```diff
@@ -0,0 +1,328 @@
+        if (!fds) {
+            fds = (struct fdsinfo64 *)malloc((size_t)FDSINFOSIZE);
+            if (!fds) {
+                PyErr_SetFromErrno(PyExc_OSError);
```
`PyErr_NoMemory()`

giampaolo commented on setup.py:
```diff
@@ -254,6 +264,8 @@ def get_ethtool_macro():
         if platform.release() == '5.10':
             posix_extension.sources.append('psutil/arch/solaris/v10/ifaddrs.c')
             posix_extension.define_macros.append(('PSUTIL_SUNOS10', 1))
+    if AIX:
```
`elif` is better

giampaolo commented:
Good work on addressing the first set of comments! I'm gonna add another "big" request.
Right now "/proc" is hard coded but in the py and the C file. It should be possible to change that via `psutil.PROCFS_PATH`. In order to do that.
Copy this in `_psaix.py`:
https://github.com/Infinidat/psutil/blob/cac89d522d64571a53011cd0914a654a9c5e2869/psutil/_pslinux.py#L215
...then store it in the `Process` class like this:
https://github.com/Infinidat/psutil/blob/cac89d522d64571a53011cd0914a654a9c5e2869/psutil/_pslinux.py#L1392
...and pass that string to the underlying C functions that need it. 

On the C side of things you'll now have to accept 2 args, the PID and /proc path prefix, like this:
https://github.com/Infinidat/psutil/blob/cac89d522d64571a53011cd0914a654a9c5e2869/psutil/_psutil_sunos.c#L96-L101

Not sure if I've been clear (I hope =)).

wiggin15 commented:
What's the reason for making `/proc` changeable? Is this configurable on the operating system?

giampaolo commented:
> What's the reason for making /proc changeable?

https://github.com/giampaolo/psutil/issues/558

wiggin15 commented:
I think, instead of moving the `get` function to `_common`, we can move this function to the main `__init__.py` file and then we won't have to use the `sys.modules` trick. Something like this in `__init__`:

    if LINUX or SOLARIS or AIX:
        # This is a public writable constant which is read from platform-specific code
        PROCFS_PATH = "/proc"

        def get_procfs_path():
            return PROCFS_PATH

What do you think?
In addition, I noticed that the BSD code is still using hard-coded `/proc`. Is it not relevant in that platform? Maybe we can change to `if POSIX` instead of Linux/Solaris/AIX if the BSD module is changed.

giampaolo commented:
How can you use that from `_ps*.py` modules if it's defined in `__init__.py`? It's `__init__` which imports those module not vice versa. Anyway it looks like something which can be dealt with in another PR so for the time being you can also define it in `_psaix.py` and then we can move it later.

> In addition, I noticed that the BSD code is still using hard-coded /proc. Is it not relevant in that platform? 

It's not, BSD systems do not rely on /proc.

wiggin15 commented:
> How can you use that from _ps*.py modules if it's defined in __init__.py?

It's a matter of import order. If you define it in `__init__` before importing the platform-specific modules, they will succeed in importing from `__init__`. Maybe not the best way to do it though.


> BSD systems do not rely on /proc.

`_psbsd.py` has a few `readlink`s and `open`s from `/proc` (the C extension doesn't, but still..)

wiggin15 commented:
Added more commits :)

giampaolo commented on psutil/TODO.aix:
```diff
@@ -0,0 +1,25 @@
+    test_procinfo               missing API
+    test_cpu_stats              cpu_stats always returns ctx_switches=0
+    test_disk_usage             df returns "-" for all procfs fields but API returns something real
+    test_pid_exists_2           there are pids in /proc that don't really exist
```
This still worries me and should be discussed. I suppose I'll have a better idea of what this means once I look at the test results.

giampaolo commented on psutil/__init__.py:
```diff
@@ -890,11 +898,12 @@ def num_handles(self):
-        switches performed by this process.
-        """
-        return self._proc.num_ctx_switches()
+    if not AIX:
```
I suggest using `if hasattr(_psplatform.Process, "num_ctx_switches"):` instead

giampaolo commented on docs/index.rst:
```diff
@@ -2110,7 +2120,7 @@ Constants
+  The path of the /proc filesystem on Linux, Solaris and AIX (defaults to
   ``"/proc"``).
   You may want to re-set this constant right after importing psutil in case
   your /proc filesystem is mounted elsewhere or if you want to retrieve
```
Please also add the new `AIX` constant in the doc and mark it as `.. versionadded:: 5.4.0`.

wiggin15 replied:
The docs have a section called `.. _const-oses:`. Is it ok if I add `AIX` to this section with:

    .. versionchanged:: 5.4.0 added AIX

in order to keep this const in the same group as the rest?

giampaolo commented on psutil/arch/aix/ifaddrs.c:
```diff
@@ -0,0 +1,125 @@
+        fam = ifr->ifr_addr.sa_family;
+
+        if (fam == AF_INET || fam == AF_INET6) {
+            cifa = (struct ifaddrs *) calloc(1, sizeof(struct ifaddrs));
```
It seems you need to check `calloc` return value

giampaolo commented on psutil/arch/aix/net_connections.c:
```diff
@@ -0,0 +1,328 @@
+    if (lseek64(Kd, (off64_t)addr, L_SET) == (off64_t)-1)
+        return(1);
+    br = read(Kd, buf, len);
+    return((br == len) ? 0 : 1);
```
If I understand this correctly you return `1` (meaning error) in case of len/size mismatch. Whereas the previous `lseek` and `read` calls set errno, this line (53) won't, so I think you need to do this:

```
if (br != len) {
    PyErr_SetString(PyExc_RuntimeError, 
                    "size mismatch when reading kernel memory fd");
    return 1;
}
return 0;
```

Also, since you set the exception in here, you should do the same for the 2 calls above (`lseek` and `read`) and do `PyErr_SetFromErrno(PyExc_OSError);`.
Of course the callers of this function should just return NULL.

giampaolo commented on psutil/arch/aix/net_connections.c:
```diff
@@ -0,0 +1,328 @@
+        &&  !psutil_kread(Kd, (KA_T)mb.m_hdr.mh_data, (char *)&un, sizeof(un))
+        ) {
+        ua = &un;
+        }
```
Can you please indent it like this (easier to read)?:

```
        if (mb.m_hdr.mh_data && !psutil_kread(
                Kd, (KA_T)mb.m_hdr.mh_data, (char *)&un, sizeof(un))) {
            ua = &un;
        }
```

giampaolo commented on psutil/arch/aix/net_connections.c:
```diff
@@ -0,0 +1,328 @@
+        }
+
+        if (unp.unp_addr) {
+            read_unp_addr(Kd, unp.unp_addr, unix_laddr_str,
```
Shouldn't you check return value? `read_unp_addr` uses `psutil_kread` so it seems you should.

giampaolo commented on psutil/arch/aix/net_connections.c:
```diff
@@ -0,0 +1,328 @@
+                PyErr_SetFromErrno(PyExc_OSError);
+                return NULL;
+            }
+            read_unp_addr(Kd, unp.unp_addr, unix_raddr_str,
```
Shouldn't you check return value? `read_unp_addr` uses `psutil_kread` so it seems you should.

giampaolo commented on psutil/arch/aix/net_connections.c:
```diff
@@ -0,0 +1,328 @@
+    if (!s.so_proto ||
+        psutil_kread(Kd, (KA_T)s.so_proto, (char *)&p, sizeof(p))) {
+        PyErr_SetFromErrno(PyExc_OSError);
+        return NULL;
```
Mmmm code smell. If `!s.so_proto` condition is true `errno` will never be set and you'll return an "empty" exception.

giampaolo commented on psutil/arch/aix/net_connections.c:
```diff
@@ -0,0 +1,328 @@
+        return NULL;
+    }
+
+    if (!p.pr_domain
```
Same here

giampaolo commented on psutil/arch/aix/net_connections.c:
```diff
@@ -0,0 +1,328 @@
+    if (fam == AF_INET || fam == AF_INET6) {
+        /* Read protocol control block */
+        if (!s.so_pcb
+        ||  psutil_kread(Kd, (KA_T) s.so_pcb, (char *) &inp, sizeof(inp))) {
```
And here

giampaolo commented on psutil/arch/aix/net_connections.c:
```diff
@@ -0,0 +1,328 @@
+            return NULL;
+        }
+        if ((KA_T) f.f_data != (KA_T) unp.unp_socket) {
+            PyErr_SetFromErrno(PyExc_OSError);
```
Mmmm I suppose `errno` will not be set here, so this will fail with an "empty" exception. Perhaps you want RuntimeError?

giampaolo commented on psutil/arch/aix/net_connections.c:
```diff
@@ -0,0 +1,328 @@
+    size_t msz;
+    pid32_t requested_pid;
+    pid32_t pid;
+    int Np = 0;          /* number of processes */
```
Why not call it  `num_procs`? It would be less confusing also because there's another `np` variable.

wiggin15 replied:
This is from lsof code, which is pretty terrible:
https://fossies.org/dox/lsof_4.89_src/aix_2dproc_8c_source.html
Interestingly both `np` and `Np` are for the number of processes in `processes`. `Np` is the number allocated and `np` is the number read. I know these are not good names but I prefer to stick to the original names for comparing to the original code (it helped just now).

giampaolo replied:
I see. OK then. And yes, I remember lsof source code: I remember I gave up because I didn't understand it. =)

giampaolo commented on psutil/arch/aix/net_connections.c:
```diff
@@ -0,0 +1,328 @@
+
+    Kd = open(KMEM, O_RDONLY, 0);
+    if (Kd < 0) {
+        PyErr_SetFromErrno(PyExc_OSError);
```
Looks like you can use `PyErr_SetFromErrnoWithFilename` instead

giampaolo commented on psutil/arch/aix/net_connections.c:
```diff
@@ -0,0 +1,328 @@
+    }
+
+    if (i > 0)
+        np += i;
```
Is this on purpose?

wiggin15 replied:
Also from [lsof code](https://fossies.org/dox/lsof_4.89_src/aix_2dproc_8c_source.html), line 476. This code is tricky but it's correct. We're reading the processes into `p` in `PROCINFO_INCR` increments. We `realloc` the next "block" for `processes` every time we read a full `PROCINFO_INCR` block. The last call can return less than `PROCINFO_INCR` processes, but we still read that block and need to go over it, so `np` must be increased by the number of processes in the last block - which is in `i`. See the [docs for `getprocs64`](https://www.ibm.com/support/knowledgecenter/en/ssw_aix_61/com.ibm.aix.basetrf1/getprocs.htm)
Confusing, I know.

wiggin15 commented:
Added a new commit and uploaded the latest test results to https://gist.github.com/wiggin15/1d76d1dc1ad50e6665fe4dc5c6fa898a
please ignore 2 tests that fail with `python2.7.bin`, this is caused by a trick we're using on our environment.

giampaolo commented:
```
  File "/root/psutil/scripts/procinfo.py", line 228, in run
    print_("ctx-switches", str_ntuple(pinfo['num_ctx_switches']))
KeyError: 'num_ctx_switches'
```

This is easy: just change scripts/procinfo.py so that it won't print this info on AIX

```
  File "/root/psutil/psutil/tests/test_posix.py", line 368, in df
    total = int(fields[1]) * 1024
ValueError: invalid literal for int() with base 10: '-'
```

This means `df -k` output is different on AIX compared to other POSIXes. You can simply skip thi test on AIX.

```
  File "/root/psutil/psutil/tests/test_contracts.py", line 541, in cwd
    self.assertIsInstance(ret, str)
AssertionError: None is not an instance of <type 'str'>
```

Just return an empty string in cwd() method of _psaix.py instead of None.

```
======================================================================
FAIL: psutil.tests.test_misc.TestMisc.test_setup_script
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/root/psutil/psutil/tests/test_misc.py", line 369, in test_setup_script
    self.assertRaises(SystemExit, module.setup)
AssertionError: SystemExit not raised
```

Just skip this one.

```
======================================================================
FAIL: psutil.tests.test_misc.TestProcessUtils.test_create_zombie_proc
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/root/psutil/psutil/tests/test_misc.py", line 950, in test_create_zombie_proc
    zpid = create_zombie_proc()
  File "/root/psutil/psutil/tests/__init__.py", line 366, in create_zombie_proc
    call_until(lambda: zproc.status(), "ret == psutil.STATUS_ZOMBIE")
  File "/root/psutil/psutil/tests/__init__.py", line 593, in wrapper
    return fun(*args, **kwargs)
  File "/root/psutil/psutil/tests/__init__.py", line 643, in call_until
    assert eval(expr)
AssertionError
```

OK, you said `create_zombie_proc()` doesn't work on AIX. Have you checked whether `ps` is able to detect it though? I just want to be sure it's not `status()` method which is broken. If we make sure it's not and we don't find a way to correctly create a zombie process via unit tests then let' just skip this. 

```

  File "/root/psutil/psutil/tests/test_process.py", line 720, in test_cmdline
    ' '.join(cmdline))
AssertionError: '/root/python/bin/python2.7 -c' != '/root/python/bin/python2.7 -c import time; time.sleep(60)'
```

I would just change the test and do a comparison of the first 30 chars if you're on AIX, else just do what we're currently doing.

```
  File "/root/psutil/psutil/tests/test_process.py", line 322, in test_io_counters
    self.assertGreater(io2.read_count, io1.read_count)
AssertionError: 0L not greater than 0L
```

Same here. If we know io_counters()'s read_count is always zero just skip this specific test line when on AIX.

```
======================================================================
FAIL: psutil.tests.test_process.TestProcess.test_prog_w_funky_name
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/root/psutil/psutil/tests/test_process.py", line 755, in test_prog_w_funky_name
    self.assertEqual(p.cmdline(), cmdline)
AssertionError: Lists differ: [''] != ['/root/psutil/$testfnfoo bar ...
```

This is bad and should be fixed. The test may also not pass but we should never have `['']` returned.

```

======================================================================
FAIL: psutil.tests.test_system.TestSystemAPIs.test_cpu_stats
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/root/psutil/psutil/tests/test_system.py", line 746, in test_cpu_stats
    self.assertGreater(value, 0)
AssertionError: 0 not greater than 0
```

Again, if we know 'ctx_switches' and 'interrupts' are always 0 just skip this test on solaris (just add a comment / note)


```
======================================================================
FAIL: psutil.tests.test_system.TestSystemAPIs.test_pid_exists_2
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/root/psutil/psutil/tests/test_system.py", line 238, in test_pid_exists_2
    self.fail(pid)
AssertionError: 260
```

This is bad. It means pid_exists() is broken. Does AIX have /proc/{pid}/status? Linux uses a specific logic for pid_exists() as it takes the Tid (thread ID) into account. I wonder if AIX does the same.


What about memory leak tests?

giampaolo commented:
Note: it may also be `pids()` returning a PID which does not exist, not necessarily `pid_exists` which is broken. Anyway, it's either one or the other. `ps` is your friend and should tell you who's wrong.

wiggin15 commented:
Added a new commit

> Just return an empty string in cwd() method of _psaix.py instead of None.

I noticed that the test for `cwd` already checks whether the return value is `None` (which can also happen on BSD and Solaris), but the test checks if it's `None` after making sure it's a string, which means it's a bug in the test. I fixed the test instead. I also fixed some other failures in `test_fetch_all`

> test_setup_script

I realized why it failed. It's because I used `python setup.py test` instead of `make test` to run the unit tests. Running the former makes argv ok for setup.setup() so it doesn't raise SystemExit with usage. I'll use `make test` from now on.

> OK, you said create_zombie_proc() doesn't work on AIX. Have you checked whether ps is able to detect it though? I just want to be sure it's not status()

OK, so it turns out you *can* create zombie processes and there was indeed a bug in `status`. It's not simple to determine actual process state from /proc, but I think I added the code required.

> I would just change the test and do a comparison of the first 30 chars if you're on AIX

The problem isn't that we cut to 30 characters (according to the struct in `/proc` the max should be 80), but that there is a single argument `import time; time.sleep(60)` which is too long so it's somehow truncated from `psinfo`. If I add another argument to argv after this argument, it will show in psinfo while "skipping" the long argument.

> It means pid_exists() is broken

The PIDs in question do exist, in `ps` and in `/proc`. They have `/proc/<pid>/psinfo`, and even `psutil.Process(pid)` works on them. Yet pid_exists returns False, because the "send a signal" check, even when trying with Python's `os.kill`, raises `OSError: no such process`. I noticed these processes are all named `wait` and it appears that there is 1 per logical CPU. It's an AIX thing...

> The test may also not pass but we should never have [''] returned.

I traced this back to an issue with our environment, again. `create_exe` used to compile C code for the test executable but now it's using `sys.executable` instead (also, there's a bug in that function where the default code is never used). Apparently `sys.executable` is not copy-able on our environment (again the `execve python2.7.bin` trick) so when running it it immediately exits, and that's why we get the empty cmdline. When changing `create_exe` (not committed, of course), those tests pass except `test_prog_w_funky_name` which, again, doesn't pass on Solaris either.

> What about memory leak tests?

Those tests pass now after a few small changes.

New test results in: https://gist.github.com/wiggin15/d1054f0a4a172a572dd610aad649dfa7

giampaolo commented:
> The PIDs in question do exist, in ps and in /proc. They have /proc/<pid>/psinfo, and even psutil.Process(pid) works on them. Yet pid_exists returns False, because the "send a signal" check, even when trying with Python's os.kill, raises OSError: no such process. I noticed these processes are all named wait and it appears that there is 1 per logical CPU. It's an AIX thing...

OK, so I suppose the `os.kill` trick is not reliable. What about `return os.path.exitsts("/proc/{pid}")`? Does it work? Last possibility is doing `return pid in pids()` but that's slow.... :-\
I will reply to other comments later (need to sleep =)).

giampaolo commented on psutil/_psutil_aix.c:
```diff
@@ -0,0 +1,870 @@
+    /* check how many perfstat_netinterface_t structures are available */
+    tot = perfstat_netinterface(NULL, NULL, sizeof(perfstat_netinterface_t), 0);
+    if (tot == 0) {
+        PyErr_SetString(PyExc_RuntimeError, "no network interface found");
```
I think it's legitimate to have a box without any NIC installed. We probably better return an empty dict.

giampaolo commented on psutil/_psutil_aix.c:
```diff
@@ -0,0 +1,870 @@
+            diskt[i].rblks * diskt[i].bsize,
+            diskt[i].wblks * diskt[i].bsize,
+            diskt[i].rserv / 1000 / 1000,  // from nano to milli secs
+            diskt[i].wserv / 1000 / 1000   // from nano to milli secs
```
Please add a XXX comment telling the field which is always set to 0

wiggin15 replied:
These values (disk_io_counters) are ok, it's `proc_io_counters` that's returning 0s. I will add `XXX` comments there.

giampaolo commented on psutil/arch/aix/ifaddrs.h:
```diff
@@ -0,0 +1,33 @@
+  unsigned int     ifa_flags;
+  struct sockaddr *ifa_addr;
+  struct sockaddr *ifa_netmask;
+  struct sockaddr *ifa_dstaddr;
```
Please indent by 4

giampaolo commented on psutil/tests/test_contracts.py:
```diff
@@ -282,6 +284,9 @@ def test_fetch_all(self):
         ])
+        if AIX:
+            # "<exiting>" processes really don't have names
+            excluded_names.add('name')
```
I wouldn't do this. Instead I would change the `name` method (defined later) like this:

```
    def name(self, ret, proc):
        self.assertIsInstance(ret, str)
        if not AIX:
            assert ret
        # ...else on AIX "<exiting>" processes (zombies) don't have names
```

giampaolo commented on psutil/tests/test_memory_leaks.py:
```diff
@@ -288,6 +289,7 @@ def test_num_fds(self):
         self.execute(self.proc.num_fds)
 
     @skip_if_linux()
+    @unittest.skipIf(AIX, "num_ctx_switches not available on AIX")
```
Please instead define a HAS_NUM_CTX_SWITCHES global in `tests/__init__.py` and use it here. I prefer that way so I can keep track of supported things in a single place, in case their support is added later.

giampaolo commented:
Also, can you please make sure all C lines are less than 80 chars long?

wiggin15 commented on psutil/_psutil_aix.c:
```diff
@@ -0,0 +1,870 @@
+        Py_DECREF(py_dev);
+        Py_DECREF(py_mountp);
+        Py_DECREF(py_tuple);
+        mt = getmntent(file);
```
This is just like the Linux and Solaris implementations and I think it's ok.

wiggin15 commented on psutil/_psaix.py:
```diff
@@ -0,0 +1,592 @@
+                return exe
+            # not found, move to search in PATH using basename only
+            exe = os.path.basename(exe)
+        # search for exe name PATH
```
There's a comment at the top of the function, "guessing is more complex than what's in the wrapping class". I don't really remember this, but it appears that we have to search in `PATH` which the parent class doesn't do.

giampaolo replied:
Yeah that's what I don't like as it seems too risky. You may have two exes with the same name in different PATHs. Plus it may be a bit too expensive / slow.

wiggin15 commented:
Added a new commit with AIX-specific tests

giampaolo commented:
As far as I'm concerned this looks good enough to be merged at this point so just tell me when you want me to do it.

wiggin15 commented:
I think that this is ready now, except I'm not sure TODO.aix should be where it is (or stay at all), and maybe you want to update the main docs about the state of support for AIX (experimental?)

giampaolo commented:
It's OK, I will update doc / README, etc. before the next release, once I'm back from China. In the meantime, thanks a lot. You really did an amazing work. At first I was skeptical about adding support for a platform I can't test against, but the quality of the code you provided is excellent and you were great at addressing all my remarks.  I hope you will keep hacking on psutil as a main contributor.
